Src/FluentAssertions/Formatting/FormattedObjectGraph.PossibleMultilineFragment.cs

```diff
@@ -0,0 +1,91 @@
+﻿using System;
+
+namespace FluentAssertions.Formatting;
+
+/// <summary>
+/// Implementation of the contained PossibleMultilineFragment.
+/// </summary>
+public partial class FormattedObjectGraph
+{
+    /// <summary>
+    /// Write fragments that may be on a single line or span multiple lines,
+    /// and this is not known until later parts of the fragment are written.
+    /// </summary>
+    internal record PossibleMultilineFragment
+    {
+        private readonly FormattedObjectGraph formattedGraph;
+        private readonly int startingLineBuilderIndex;
+        private readonly int startingLineCount;
+
+        public PossibleMultilineFragment(FormattedObjectGraph formattedGraph)
+        {
+            this.formattedGraph = formattedGraph;
+            startingLineBuilderIndex = formattedGraph.lineBuilder.Length;
+            startingLineCount = formattedGraph.lines.Count;
+        }
+
+        /// <summary>
+        /// Write the fragment at the position the graph was in when this instance was created.
+        ///
+        /// <para>
+        /// If more lines have been added since this instance was created then write the
+        /// fragment on a new line, otherwise write it on the same line.
+        /// </para>
+        /// </summary>
+        internal void AddFragmentAtStart(string fragment)
+        {
+            if (FormatOnSingleLine)
+            {
+                formattedGraph.lineBuilder.Insert(startingLineBuilderIndex, fragment);
+            }
+            else
+            {
+                formattedGraph.lines.Insert(startingLineCount, Environment.NewLine + fragment);
+                InsertAtStartOfLine(startingLineCount + 1, MakeWhitespace(1));
+            }
+        }
+
+        private bool FormatOnSingleLine => formattedGraph.lines.Count == startingLineCount;
+
+        private void InsertAtStartOfLine(int lineIndex, string insertion)
+        {
+            formattedGraph.lines[lineIndex] = formattedGraph.lines[lineIndex].Insert(0, insertion);
+        }
+
+        /// <summary>
+        /// If more lines have been added since this instance was created then write the
+        /// fragment on a new line, otherwise write it on the same line.
+        /// </summary>
+        internal void AddLineOrFragment(string fragment)
```

Comment:
🤔 Since the other methods also do not specify in their name whether the fragment will end-up on a separate line, `AddFragment` would be a better fit.

Reply:
There is already an `AddFragment` that specifically *doesn't* add a newline even if the output is multi-line.

How about

    AddStartingLineOrFragment
    AddLineOrFragment
    AddFragment
    AddEndingLineOrFragment

??

Reply:
Oh, I missed that one. Well, then just keep it as it.